staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -432,3 +443,42 @@ func (r *Reflector) setLastSyncResourceVersion(v string) {
 	defer r.lastSyncResourceVersionMutex.Unlock()
 	r.lastSyncResourceVersion = v
 }
+
+// relistResourceVersion determines the resource version the reflector should list or relist from.
+// Returns either the lastSyncResourceVersion so that this reflector will relist with a resource
+// versions no older than has already been observed in relist results or watch events, or, if the last relist resulted
+// in an HTTP 410 (Gone) status code, returns "" so that the relist will use the latest resource version available in
+// etcd via a quorum read.
+func (r *Reflector) relistResourceVersion() string {
+	r.lastSyncResourceVersionMutex.RLock()
+	defer r.lastSyncResourceVersionMutex.RUnlock()
+
+	if r.isLastSyncResourceVersionGone {
+		// Since this reflector makes paginated list requests, and all paginated list requests skip the watch cache
+		// if the lastSyncResourceVersion is expired, we set ResourceVersion="" and list again to re-establish reflector
+		// to the latest available ResourceVersion, using a consistent read from etcd.
+		return ""
+	}
+	if r.lastSyncResourceVersion == "" {
+		// For performance reasons, initial list performed by reflector uses "0" as resource version to allow it to
+		// be served from the watch cache if it is enabled.
+		return "0"
+	}
+	return r.lastSyncResourceVersion
+}
+
+// setIsLastSyncResourceVersionGone sets if the last list or watch request with lastSyncResourceVersion returned a
+// 410 (Gone).
+func (r *Reflector) setIsLastSyncResourceVersionGone(isExpired bool) {
+	r.lastSyncResourceVersionMutex.Lock()
+	defer r.lastSyncResourceVersionMutex.Unlock()
+	r.isLastSyncResourceVersionGone = isExpired
+}
+
+func isGoneError(err error) bool {
```

Comment:
wait, we don't have an errors helper for this?  Would rather add one.

Reply:
I'm really confused why we need the helper vs `apierrs.IsResourceExpired`?

Reply:
`apierrs.IsResourceExpired` only matches status code `410` where `reason=apierrs.StatusReasonExpired`. But we also have `apierrs.StatusReasonGone` for status code `410`. We must handle both here.

Reply:
I'll explain in the comment too.

Reply:
I'm okay with putting a helper in as well. The complication with that is that there is already a `apierrs.IsResourceGone` function that only matches `apierrs.StatusReasonGone` and not `apierrs.StatusReasonExpired`

Reply:
Why would the server respond with 410 but not cause expired?  We can have multiple causes, so any place where we expire on server we should set cause expired.

Reply:
Put another way, no server code indicating “resource version too old” should return 410 without returning reason expired.  If we have older versions of the server that don’t do that, we should have two checks and a comment on when to remove this check (Ie whenever client lib should stop responding).

Reply:
Here's where the server returns a 410 using `errors.NewGone`: https://github.com/kubernetes/kubernetes/blob/c5170dfb66749b60a4a1611dc7b4a39e64b33e2e/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L471. I'll update that to return `errors.NewResourceExpired` and since that was the last remaining use of `NewGone`, I've marked it as deprecated.

I've swtiched to using two checks and have added a comment explaining why they are both needed and the conditions for removing the one we are deprecating.